File: src/minigames/pillar_battle/java/work/lclpnet/ap2/game/pillar_battle/item/EnchantedBookItemClass.java
package work.lclpnet.ap2.game.pillar_battle.item;

import net.minecraft.core.RegistryAccess;
import net.minecraft.core.component.DataComponents;
import net.minecraft.core.registries.Registries;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.Items;
import net.minecraft.world.item.enchantment.Enchantment;
import net.minecraft.world.item.enchantment.EnchantmentHelper;
import net.minecraft.world.item.enchantment.ItemEnchantments;
import work.lclpnet.ap2.impl.util.ItemHelper;

import java.util.Random;
import java.util.stream.Stream;

public class EnchantedBookItemClass implements ItemClass {

    private final RegistryAccess registryManager;

    public EnchantedBookItemClass(RegistryAccess registryManager) {
        this.registryManager = registryManager;
    }

    @Override
    public ItemStack getRandomStack(Random random) {
        ItemStack stack = new ItemStack(Items.ENCHANTED_BOOK);

        var registry = registryManager.lookupOrThrow(Registries.ENCHANTMENT);
        var entry = ItemHelper.getRandomEntry(registry, random);

        if (entry == null) return stack;

        Enchantment enchantment = entry.value();

        int minLevel = enchantment.getMinLevel();
        int level = minLevel + random.nextInt(enchantment.getMaxLevel() - minLevel + 1);

        var builder = new ItemEnchantments.Mutable(EnchantmentHelper.getEnchantmentsForCrafting(stack));
        builder.set(entry, level);

        stack.set(DataComponents.STORED_ENCHANTMENTS, builder.toImmutable());

        return stack;
    }

    @Override
    public Stream<Item> stream() {
        return Stream.of(Items.ENCHANTED_BOOK);
    }
}

File: src/minigames/pillar_battle/src/main/kotlin/work/lclpnet/ap2/game/pillar_battle/item/EnchantedBookItemClass.kt
package work.lclpnet.ap2.game.pillar_battle.item

import net.minecraft.core.RegistryAccess
import net.minecraft.core.component.DataComponents
import net.minecraft.core.registries.Registries
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.Items
import net.minecraft.world.item.enchantment.EnchantmentHelper
import net.minecraft.world.item.enchantment.ItemEnchantments
import work.lclpnet.ap2.impl.util.ItemHelper
import java.util.*
import java.util.stream.Stream

class EnchantedBookItemClass(private val registryManager: RegistryAccess) : ItemClass {
    override fun getRandomStack(random: Random): ItemStack {
        val stack = ItemStack(Items.ENCHANTED_BOOK)

        val registry = registryManager.lookupOrThrow(Registries.ENCHANTMENT)
        val entry = ItemHelper.getRandomEntry(registry, random) ?: return stack

        val enchantment = entry.value()
        val minLevel = enchantment.minLevel
        val level = minLevel + random.nextInt(enchantment.maxLevel - minLevel + 1)

        val builder = ItemEnchantments.Mutable(EnchantmentHelper.getEnchantmentsForCrafting(stack))
        builder.set(entry, level)

        stack.set(DataComponents.STORED_ENCHANTMENTS, builder.toImmutable())
        return stack
    }

    override fun stream(): Stream<Item> = Stream.of(Items.ENCHANTED_BOOK)
}
